CrashTracer: com.apple.WebKit.GPU at com.apple.WebCore: WebCore::RangeResponseGenerator::removeTask

youssefsoli opened the pull request:
#### fd72cc6f23e9c0bf6834183336d0e19b23127e3d
<pre>
CrashTracer: com.apple.WebKit.GPU at com.apple.WebCore: WebCore::RangeResponseGenerator::removeTask
<a href="https://bugs.webkit.org/show_bug.cgi?id=243593">https://bugs.webkit.org/show_bug.cgi?id=243593</a>
&lt;rdar://77013610&gt;

Reviewed by Chris Dumez.

Our HashMap implementation treats null Strings as a special value and
when such a string is passed onto `HashMap::get()` we end up dereferencing
a nullptr within the String implementation.

The fix is to null check the string we pass into `HashMap::get()`.

* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::removeTask):
(WebCore::RangeResponseGenerator::willHandleRequest):

Canonical link: <a href="https://commits.webkit.org/253164@main">https://commits.webkit.org/253164@main</a>
</pre>

cdumez commented on Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
```diff
@@ -96,7 +96,11 @@ static ResourceResponse synthesizedResponseForRange(const ResourceResponse& orig
-    auto* data = m_map.get(task.originalRequest.URL.absoluteString);
+    auto urlString = task.originalRequest.URL.absoluteString;
+    // HashMap::get() crashes if a null String is passed.
+    if (!urlString)
```
This doesn't look great. We may end up leaving tasks in m_map that we don't want to.

Why do we end up with a null String here? How come were were able to add the task to m_map in the first place (if its urlString was originally null)?

jernoble replied:
CC @achristensen07

cdumez replied:
`RangeResponseGenerator::willSynthesizeRangeResponses()` calls:
```
 m_map.ensure(originalRequest.URL.absoluteString, [&] {
```

with `originalRequest` being `task.originalRequest`.

So if `task.originalRequest.URL.absoluteString` was nil, we would have crashed on adding the task to the map it seems. Something else must be going on here.

cdumez replied:
Oh, `RangeResponseGenerator::willSynthesizeRangeResponses()` does have this check:
```
    if (!originalRequest.URL)
        return false;
```

So the issue may be that we create the task (with a null URL for some reason), we call willSynthesizeRangeResponses() which will return early and not add the task to m_map. Then we cancel the task, which calls `removeTask()` but we don't have the same null URL check there.

Given that, I think the patch is OK. The task wouldn't be in m_map anyway.

cdumez commented on Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
```diff
@@ -96,7 +96,11 @@ static ResourceResponse synthesizedResponseForRange(const ResourceResponse& orig
 {
     ASSERT(isMainThread());
-    auto* data = m_map.get(task.originalRequest.URL.absoluteString);
+    auto urlString = task.originalRequest.URL.absoluteString;
```
Maybe we should do the same early return like the function that is adding the task:
```
if (!originalRequest.URL)
    return;
```

youssefsoli replied:
Makes sense.

cdumez reviewed (approved):
LGTM

youssefsoli committed:
CrashTracer: com.apple.WebKit.GPU at com.apple.WebCore: WebCore::RangeResponseGenerator::removeTask
https://bugs.webkit.org/show_bug.cgi?id=243593
<rdar://77013610>

Reviewed by Chris Dumez.

Our HashMap implementation treats null Strings as a special value and
when such a string is passed onto `HashMap::get()` we end up dereferencing
a nullptr within the String implementation.

The fix is to null check the string we pass into `HashMap::get()`.

* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::removeTask):
(WebCore::RangeResponseGenerator::willHandleRequest):

Canonical link: https://commits.webkit.org/253164@main
